cpu/stm32/periph/eth.c

```diff
@@ -615,22 +627,23 @@ static int stm32_eth_recv(netdev_t *netdev, void *buf, size_t max_len,
         DEBUG("[stm32_eth] Buffer provided by upper layer is too small\n");
         drop_frame_and_update_rx_curr();
         return -ENOBUFS;
-    }
+   }
 
-    size_t remain = size;
-    while (remain) {
-        size_t chunk = MIN(remain, ETH_RX_BUFFER_SIZE);
-        memcpy(data, rx_curr->buffer_addr, chunk);
-        data += chunk;
-        remain -= chunk;
-        /* Hand over descriptor to DMA */
-        rx_curr->status = RX_DESC_STAT_OWN;
-        rx_curr = rx_curr->desc_next;
-    }
+    copy_from_dma_desc(buf, size, rx_curr);
 
-    if ((size + ETHERNET_FCS_LEN - 1) % ETH_RX_BUFFER_SIZE < ETHERNET_FCS_LEN) {
-        /* one additional rx descriptor was needed only for the FCS, hand that
-         * back to the DMA as well */
+    /* Hand over DMA descriptors back to DMA, collect RX timestamp from last
+     * descriptor if module periph_ptp is used */
+    while (1) {
+        if (rx_curr->status & RX_DESC_STAT_LS) {
+            if (IS_USED(MODULE_PERIPH_PTP)) {
+                info->timestamp = rx_curr->ts_low;
+                info->timestamp += (uint64_t)rx_curr->ts_high * NS_PER_SEC;
+                info->flags |= NETDEV_ETH_RX_INFO_FLAG_TIMESTAMP;
+            }
+            rx_curr->status = RX_DESC_STAT_OWN;
+            rx_curr = rx_curr->desc_next;
+            break;
+        }
         rx_curr->status = RX_DESC_STAT_OWN;
         rx_curr = rx_curr->desc_next;
```

Comment:
Couldn't this also be done in `copy_from_dma_desc()` so we don't have to iterate the list a second time?

Reply:
It is not so trivial. There are some special cases:

- When the CRC sum didn't fit in the last descriptor used for the Ethernet SDU, one additional descriptor is used to store the (remaining bytes of the) CRC sum. Hence, one descriptor more needs to be freed than was copied
- The timestamp needs to be taken from the last descriptor *before* it is passed on to the DMA

It would however be nice to free each descriptor after the data is fetched, as this would result in better handling of traffic bursts. (The DMA sooner gets the descriptors back, which decreases the chance of frames being dropped.) I will give it a second look on whether I manage to implement this more elegant.

Reply:
It is now a single loop, but there is now an `if` to handle each iteration. But keeping in mind that there are usually very few (if even more than one) iteration of the loop, this is little value in optimizing the loop for low cycle count anyway.